File: master-node-docker/sentinel/eth/vpn_contract.py
# coding=utf-8
import rlp
from ethereum.transactions import Transaction

from .eth import eth_manager
from ..config import COINBASE_ADDRESS
from ..config import COINBASE_PRIVATE_KEY
from ..config import VPN_SERVICE
from ..helpers import redis_manager


class VpnServiceManager(object):
    def __init__(self, net, name, address, abi):
        self.net = net
        self.address = address
        self.contract = net.web3.eth.contract(contract_name=name, abi=abi, address=address)

    def pay_vpn_session(self, account_addr, amount, session_id):
        try:
            nonce = redis_manager.get_nonce(COINBASE_ADDRESS, self.net.chain)
            tx = Transaction(nonce=nonce,
                             gasprice=self.net.web3.eth.gasPrice,
                             startgas=1000000,
                             to=self.address,
                             value=0,
                             data=self.net.web3.toBytes(hexstr=self.contract.encodeABI(fn_name='payVpnSession',
                                                                                       args=[account_addr, amount,
                                                                                             session_id])))
            tx.sign(COINBASE_PRIVATE_KEY)
            raw_tx = self.net.web3.toHex(rlp.encode(tx))
            tx_hash = self.net.web3.eth.sendRawTransaction(raw_tx)
            redis_manager.set_nonce(COINBASE_ADDRESS, self.net.chain, nonce + 1)
        except Exception as err:
            return {
                       'code': 301,
                       'error': str(err)
                   }, None
        return None, tx_hash

    def set_initial_payment(self, account_addr, is_paid=True):
        try:
            nonce = redis_manager.get_nonce(COINBASE_ADDRESS, self.net.chain)
            tx = Transaction(nonce=nonce,
                             gasprice=self.net.web3.eth.gasPrice,
                             startgas=1000000,
                             to=self.address,
                             value=0,
                             data=self.net.web3.toBytes(
                                 hexstr=self.contract.encodeABI(fn_name='setInitialPaymentStatusOf',
                                                                args=[account_addr, is_paid])))
            tx.sign(COINBASE_PRIVATE_KEY)
            raw_tx = self.net.web3.toHex(rlp.encode(tx))
            tx_hash = self.net.web3.eth.sendRawTransaction(raw_tx)
            redis_manager.set_nonce(COINBASE_ADDRESS, self.net.chain, nonce + 1)
        except Exception as err:
            return {
                       'code': 302,
                       'error': str(err)
                   }, None
        return None, tx_hash

    def get_due_amount(self, account_addr):
        try:
            caller_object = {
                'from': account_addr,
                'to': self.address,
                'data': self.net.web3.toHex(self.net.web3.toBytes(
                    hexstr=self.contract.encodeABI(fn_name='getDueAmountOf', args=[account_addr])))
            }
            due_amount = self.net.web3.toInt(hexstr=self.net.web3.eth.call(caller_object))
        except Exception as err:
            return {
                       'code': 303,
                       'error': str(err)
                   }, None
        return None, due_amount

    def get_vpn_sessions_count(self, account_addr):
        try:
            caller_object = {
                'from': account_addr,
                'to': self.address,
                'data': self.net.web3.toHex(self.net.web3.toBytes(
                    hexstr=self.contract.encodeABI(fn_name='getVpnSessionsCountOf', args=[account_addr])))
            }
            sessions_count = self.net.web3.toInt(hexstr=self.net.web3.eth.call(caller_object))
        except Exception as err:
            return {
                       'code': 304,
                       'error': str(err)
                   }, None
        return None, sessions_count

    def get_initial_payment(self, account_addr):
        try:
            caller_object = {
                'from': account_addr,
                'to': self.address,
                'data': self.net.web3.toHex(self.net.web3.toBytes(
                    hexstr=self.contract.encodeABI(fn_name='getInitialPaymentStatusOf', args=[account_addr])))
            }
            is_paid = self.net.web3.toInt(hexstr=self.net.web3.eth.call(caller_object))
        except Exception as err:
            return {
                       'code': 305,
                       'error': str(err)
                   }, None
        return None, is_paid == 1

    def get_vpn_usage(self, account_addr, session_id):
        try:
            caller_object = {
                'from': account_addr,
                'to': self.address,
                'data': self.net.web3.toHex(self.net.web3.toBytes(
                    hexstr=self.contract.encodeABI(fn_name='getVpnUsageOf', args=[account_addr, session_id])))
            }
            usage = self.net.web3.eth.call(caller_object)[2:]
            usage = [usage[i:i + 64] for i in range(0, len(usage), 64)]
            usage[0] = self.net.web3.toChecksumAddress(usage[0])
            usage[1:] = [self.net.web3.toInt(hexstr=usage[i]) for i in range(1, len(usage))]
            usage[-1] = usage[-1] != 0
        except Exception as err:
            return {
                       'code': 306,
                       'error': str(err)
                   }, None
        return None, usage

    def add_vpn_usage(self, from_addr, to_addr, sent_bytes, session_duration, amount, timestamp, session_id):
        try:
            nonce = redis_manager.get_nonce(COINBASE_ADDRESS, self.net.chain)
            tx = Transaction(nonce=nonce,
                             gasprice=self.net.web3.eth.gasPrice,
                             startgas=1000000,
                             to=self.address,
                             value=0,
                             data=self.net.web3.toBytes(hexstr=self.contract.encodeABI(fn_name='addVpnUsage',
                                                                                       args=[from_addr, to_addr,
                                                                                             sent_bytes,
                                                                                             session_duration, amount,
                                                                                             timestamp, session_id])))
            tx.sign(COINBASE_PRIVATE_KEY)
            raw_tx = self.net.web3.toHex(rlp.encode(tx))
            tx_hash = self.net.web3.eth.sendRawTransaction(raw_tx)
            redis_manager.set_nonce(COINBASE_ADDRESS, self.net.chain, nonce + 1)
        except Exception as err:
            return {
                       'code': 307,
                       'error': str(err)
                   }, None
        return None, tx_hash


vpn_service_manager = VpnServiceManager(eth_manager['rinkeby'], VPN_SERVICE['name'], VPN_SERVICE['address'],
                                        VPN_SERVICE['abi'])

File: master-node-nodejs/src/eth/vpn_contract.js
import Tx from 'ethereumjs-tx';

import { Eth_manager } from './eth'
import { VPN_SERVICE } from '../config/services';
import { ADDRESS as COINBASE_ADDRESS, PRIVATE_KEY as COINBASE_PRIVATE_KEY } from '../config/eth';

class VpnService_Manager {
  constructor(net, name, address, abi) {
    this.net = net;
    this.address = address;
    this.contract = net.web3.eth.contract(abi).at(address);
  }
  async payVpnSession(accountAddr, amount, sessionId, nonce, cb) {
    let rawTx = {
      nonce: nonce,
      gasPrice: this.net.web3.toHex(this.net.web3.eth.gasPrice),
      gasLimit: this.net.web3.toHex(500000),
      to: this.address,
      value: '0x0',
      data: this.contract.payVpnSession.getData(accountAddr, amount, sessionId)
    };
    let tx = new Tx(rawTx);
    tx.sign(Buffer.from(COINBASE_PRIVATE_KEY, 'hex'));
    let serializedTx = tx.serialize();
    this.net.web3.eth.sendRawTransaction('0x' + serializedTx.toString('hex'), (err, txHash) => {
      cb(err, txHash);
    });
  }
  setInitialPayment(accountAddr, nonce, isPayed = true) {
    let rawTx = {
      nonce: nonce,
      gasPrice: this.net.web3.toHex(this.net.web3.eth.gasPrice),
      gasLimit: this.net.web3.toHex(500000),
      to: this.address,
      value: '0x0',
      data: this.contract.setInitialPaymentOf.getData(accountAddr, isPayed)
    };
    let tx = new Tx(rawTx);
    tx.sign(Buffer.from(COINBASE_PRIVATE_KEY, 'hex'));
    let serializedTx = tx.serialize();
    this.net.web3.eth.sendRawTransaction('0x' + serializedTx.toString('hex'), (err, txHash) => {
      cb(err, txHash);
    });
  }
  getDueAmount(accountAddr, cb) {
    this.contract.getDueAmountOf(accountAddr, { from: COINBASE_ADDRESS }, (err, rawDueAmount) => {
      let dueAmount = parseInt(rawDueAmount);
      dueAmount = dueAmount / Math.pow(10, 18);
      cb(err, dueAmount);
    });
  }
  getVpnSessionCount(accountAddr, cb) {
    this.contract.getVpnSessionsCountOf(accountAddr, { from: COINBASE_ADDRESS }, (err, rawSessions) => {
      let sessions = parseInt(rawSessions);
      cb(err, sessions);
    });
  }
  getInitialPayment(account_addr, cb) {
    this.contract.getInitialPaymentStatusOf(account_addr, { from: COINBASE_ADDRESS }, (err, isPayed) => {
      cb(null, isPayed);
    });
  }
  getVpnUsage(accountAddr, index, cb) {
    this.contract.getVpnUsageOf(accountAddr, index, { from: COINBASE_ADDRESS }, (err, usage) => {
      cb(err, usage);
    });
  }
  addVpnUsage(fromAddr, toAddr, sentBytes, sessionDuration, amount, timeStamp, sessionId, nonce, cb) {
    try {
      let rawTx = {
        nonce: nonce,
        gasPrice: this.net.web3.toHex(this.net.web3.eth.gasPrice),
        gasLimit: this.net.web3.toHex(500000),
        to: this.address,
        value: '0x0',
        data: this.contract.addVpnUsage.getData(fromAddr, toAddr, sentBytes, sessionDuration, amount, timeStamp, sessionId)
      };
      let tx = new Tx(rawTx);
      tx.sign(Buffer.from(COINBASE_PRIVATE_KEY, 'hex'));
      let serializedTx = tx.serialize();
      this.net.web3.eth.sendRawTransaction('0x' + serializedTx.toString('hex'), (err, txHash) => {
        if (err)
          cb(err, null);
        else
          cb(null, txHash);
      });
    }
    catch (error) {
      cb(error, null);
    }
  }
}

export const VpnServiceManager = new VpnService_Manager(Eth_manager['rinkeby'], VPN_SERVICE['name'], VPN_SERVICE['address'], VPN_SERVICE['abi']);